kubectl/upload/util.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func homeDir() string {
+	if h := os.Getenv("HOME"); h != "" {
+		return h
+	}
+	return os.Getenv("USERPROFILE") // windows
+}
+
+func calculateMD5(path string) (string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
+}
+
+func tarGz(src, dst string) error {
+	tarFile, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer tarFile.Close()
+
+	gzWriter := gzip.NewWriter(tarFile)
+	defer gzWriter.Close()
+
+	tarWriter := tar.NewWriter(gzWriter)
+	defer tarWriter.Close()
+
+	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip the root directory
+		if path == src {
+			return nil
+		}
+
+		header, err := tar.FileInfoHeader(info, info.Name())
+		if err != nil {
+			return err
+		}
+
+		// Use relative filepath to ensure the root directory is not included in tarball
+		relativePath, err := filepath.Rel(src, path)
+		if err != nil {
+			return err
+		}
+
+		// clean up the file name to avoid including preceding "./" or "/"
+		header.Name = strings.TrimPrefix(relativePath, string(filepath.Separator))
+
+		// Skip if it is not a regular file or a directory
+		if !info.Mode().IsRegular() && !info.IsDir() {
+			return nil
+		}
+
+		if err := tarWriter.WriteHeader(header); err != nil {
+			return err
+		}
+
+		if info.Mode().IsRegular() {
+			file, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
```

Comment:
I think if this fails the write to the file will fail and the error will tossed into the ether.

Reply:
I'll add some output to make that more visibile.